### PR TITLE
Document Client Flow

### PR DESCRIPTION
**Pull request by tkz00**
## Motivation

No information about the data flows from inside the client existed, this are helpful for new and current developers to have a faster way to grab how the application works.

## Summary of changes

Added some paragraphs describing two flows from the client and diagrams to help visualize this information.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -106,3 +107,70 @@ We share our development and creative process in the open, follow us for frequen
...
 - **Telegram:** [t.me/curseofmirra](https://t.me/curseofmirra)
+
+## Documentation
+### Client Paths
+
+On the start of the application, the SocketConnection, which is a game object in the scene, tries to establish a websocket connection with the backend. If and when the connection is opened, SocketConnection asks the backend for the current user (the register/login logic is not shown in this diagram).
+
+In parallel, the HeaderManager waits for the user information to be retrived via the GlobalUserData, which is the class that contains the user information.
+
+Once the user is retreived from the backend, it is sent to the HeaderManger, where the relevant information is displayed in the UI. Also the HeaderManager subscribes to changes from the GlobalUserData, to receive the new values and update the ones on the UI.
```
Comment (ncontinanza) [inline README.md:118]:
```suggestion
The following diagram outlines the initialization process of a connection between the client and the backend, as well as the retrieval of the user's data to be displayed on the header.

On the start of the application, the SocketConnection, which is a game object in the scene, tries to establish a websocket connection with the backend. Upon successful establishment of the connection, the SocketConnection requests the current user data from the backend (note: the register/login logic is not shown in this diagram).

In parallel, the HeaderManager waits for the user information to be retrieved via the GlobalUserData, which is the class that contains the user information.

Once the user data is retrieved from the backend, it is sent to the HeaderManager, where the relevant information is displayed in the UI. Additionally, the HeaderManager subscribes to updates from the GlobalUserData, ensuring that any changes are promptly reflected in the UI. 
```

File: README.md [hunk 2]
```diff
@@ -106,3 +107,70 @@ We share our development and creative process in the open, follow us for frequen
...
+			GlobalUserData-->>-HeaderManager: User change
+			HeaderManager->>UI: User Data to Display
+		end
+```
+
+___
+
+When entering the Summon scene, the SummonManager script is responsible to retrive the summon boxes from the backend through SocketConnection, when these are returned, the manager instantiates an UI element for each box.
+
+The user then can buy one of these boxes, causing a chain comunication all the way to the backend, passing through currency availability checks first. Once the backend processes the request, it returns the summoned unit so the summon manager can display it in the UI.
```
Comment (ncontinanza) [inline README.md:152]:
```suggestion
The next diagram depicts the actions that take place when a user enters the Summon scene.

When entering the Summon scene, the SummonManager script is responsible of fetching the summon boxes from the backend via the SocketConnection. When these are returned, the manager instantiates a UI element for each box.

The user can then buy one of these boxes, triggering a chain communication that traverses all the way to the backend, passing through currency availability checks first. Once the backend processes the user's request, it returns the summoned unit so the summon manager can display it in the UI.
```

**Review by ncontinanza — CHANGES_REQUESTED** [review]
Some suggestions on the text clarity

File: README.md [hunk 3]
```diff
@@ -14,6 +14,7 @@
   - [Suggested Development Environment](#suggested-development-environment)
   - [Useful Commands](#useful-commands)
   - [Contact and Socials](#contact-and-socials)
+  - [Tecnical Documentation](#documentation)
```
Comment (ncontinanza) [inline README.md:17]:
```suggestion
  - [Technical Documentation](#documentation)
```

**Review by ncontinanza — CHANGES_REQUESTED** [review]

**Review by ncontinanza — APPROVED** [review]
LGTM!

**Review by mgrunwaldt — CHANGES_REQUESTED** [review]
The documentation is very complete and easy to follow. Having that said, there are some details in the diagram that need to be corrected:

1. In the first diagram, who or what triggers socket connection to initialize the conection? I would add something like a user opening the app
2. HeaderManager, GlobalUserData, SocketConnection, etc etc are all classes, you should specify that you are dealing with instances of that class. It's easy as adding a : before the class name ( :HeaderManager for example)
3. When sending a message (arrows to the right) you should include the name+parameters (if any) of the method / function. In some cases you have for example GetUser, it should be GetUser() (or if it sends any parameter also add it there), but on others you just describe the method (for example "Subscribe to user changes") this makes it harder to go from the diagram to the code
4. When you are returning something (arrow to the left) I prefer if we return it in camelCase, because if you send it in PascalCase it might be confusing and people might think you are returning a whole class (this is not something mandatory, just a personal preference)
5. What is the UI line? is UI a class?
6. Not sure I understand what the "par" in the first diagram represents

I reccomend taking a quick look at IMB's documentation for sequence diagrams, all the details I expressed are solved there: https://developer.ibm.com/articles/the-sequence-diagram/
 
Let's be clear, this is a very good work and all the things I just sent are tiny formalities that I'd prefer to address before we merge this PR, but the actual content of the documentation and the diagram is perfect.

**Comment by mgrunwaldt** [comment]
> > * In the first diagram, who or what triggers socket connection to initialize the conection? I would add something like a user opening the app
> > * HeaderManager, GlobalUserData, SocketConnection, etc etc are all classes, you should specify that you are dealing with instances of that class. It's easy as adding a : before the class name ( :HeaderManager for example)
> > * When sending a message (arrows to the right) you should include the name+parameters (if any) of the method / function. In some cases you have for example GetUser, it should be GetUser() (or if it sends any parameter also add it there), but on others you just describe the method (for example "Subscribe to user changes") this makes it harder to go from the diagram to the code
> > * When you are returning something (arrow to the left) I prefer if we return it in camelCase, because if you send it in PascalCase it might be confusing and people might think you are returning a whole class (this is not something mandatory, just a personal preference)
> > * What is the UI line? is UI a class?
> > * Not sure I understand what the "par" in the first diagram represents
> 
> 1. Not sure I understand what the "par" in the first diagram represents
> 
> _The “par” block means those message blocks run in [parallel](https://developer.ibm.com/articles/the-sequence-diagram/#parallel)._
> 
> 2. In the first diagram, who or what triggers socket connection to initialize the connection? I would add something like a user opening the app
> 
> _A User actor could exist, and start the application sending a message to HeaderManager and SocketConnection, but in reality the User actor doesn’t communicate with these objects directly, he just starts the application and with this event the objects start sending messages on their own._
> 
> 3. HeaderManager, GlobalUserData, SocketConnection, etc etc are all classes, you should specify that you are dealing with instances of that class. It's easy as adding a : before the class name ( :HeaderManager for example)
> 
> _Right! I will add those._
> 
> 4. What is the UI line? is UI a class?
> 
> _The UI line is a set of 5 UI objects (but that number is bound to change), they could be represented as different objects, such as (nametag : TMP_Text) or (XPBar : ProgressBar) but I believe it will cause the diagram to be confusing and less readable._
> 
> 5. When sending a message (arrows to the right) you should include the name+parameters (if any) of the method / function. In some cases you have for example GetUser, it should be GetUser() (or if it sends any parameter also add it there), but on others you just describe the method (for example "Subscribe to user changes") this makes it harder to go from the diagram to the code
> 
> _I’ll add the parameters to the method calls, but some messages, such as “User Data to Display” are a reduced representation of what the message is doing, that specific message is actually multiple messages to different UI elements (nametag, resource counters, level indicator, experience bar) updating their values, I thought it was clearer this way._
> 
> 6. When you are returning something (arrow to the left) I prefer if we return it in camelCase, because if you send it in PascalCase it might be confusing and people might think you are returning a whole class (this is not something mandatory, just a personal preference)
> 
> _Ok, great! Will correct that too_

1. Perfect
2. Don't Agree. "he just starts the application and with this event the objects start sending messages on their own". This needs to be shown somehow. When I see the diagram I need to see what triggers everything. Plus, I never saw a diagram in which the starting action is not in the line on the left. Not 100% sure this is mandatory though, only saying I never saw one like this, do you have any example or document that says that you can skip this?
3. Perfect
4. Perfect, makes sense
5. Perfect, in those cases it makes sense what you say
6. Perfect

**Review by mgrunwaldt — APPROVED** [review]
Nice!

**Review by ncontinanza — APPROVED** [review]
Nice job!

**Pull request merged**